core/server/master/src/main/java/alluxio/master/metastore/caching/Cache.java

```diff
@@ -0,0 +1,436 @@
+    return mMap.size() >= mLowWaterMark;
+  }
+
+  private boolean overHighWaterMark() {
+    return mMap.size() >= mHighWaterMark;
+  }
+
+  private boolean cacheIsFull() {
+    return mMap.size() >= mMaxSize;
+  }
+
+  private void wakeEvictionThreadIfNecessary() {
+    if (mEvictionThread.mIsSleeping && mMap.size() >= mHighWaterMark) {
+      kickEvictionThread();
+    }
+  }
+
+  private void kickEvictionThread() {
+    synchronized (mEvictionThread) {
+      if (mEvictionThread.getState() == State.NEW) {
+        mEvictionThread.start();
+      }
+      mEvictionThread.notifyAll();
+    }
+  }
+
+  @Override
+  public void close() {
+    mEvictionThread.interrupt();
+    try {
+      mEvictionThread.join(10 * Constants.SECOND_MS);
+      if (mEvictionThread.isAlive()) {
+        LOG.warn("Failed to stop eviction thread");
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    }
+  }
+
+  private class EvictionThread extends Thread {
+    private final TemporalAmount mWarnInterval = Duration.ofSeconds(30);
+
+    private volatile boolean mIsSleeping = true;
+
+    private Iterator<Entry> mEvictionHead = Collections.emptyIterator();
+    private Instant mNextAllowedSizeWarning = Instant.EPOCH;
+
+    // This is used temporarily in each call to evictEntries. We store it as a field to avoid
+    // re-allocating the array on each eviction.
+    private List<Entry> mEvictionCandidates;
+
+    private EvictionThread() {
+      super(mName + "-eviction-thread");
+      mEvictionCandidates = new ArrayList<>(mEvictBatchSize);
+    }
+
+    @Override
+    public void run() {
+      while (!Thread.interrupted()) {
+        // Wait for the cache to get over the high water mark.
+        while (!overHighWaterMark()) {
+          synchronized (mEvictionThread) {
+            if (!overHighWaterMark()) {
+              try {
+                mIsSleeping = true;
+                mEvictionThread.wait();
+                mIsSleeping = false;
+              } catch (InterruptedException e) {
+                return;
+              }
+            }
+          }
+        }
+        evictToLowWaterMark();
+      }
+    }
+
+    private void evictToLowWaterMark() {
+      Instant evictionStart = Instant.now();
+      int evictionCount = 0;
+      while (overLowWaterMark()) {
+        if (cacheIsFull()) {
+          Instant now = Instant.now();
+          if (now.isAfter(mNextAllowedSizeWarning)) {
+            LOG.warn(
+                "Metastore {} cache is full. Consider increasing the cache size or lowering the "
+                    + "high water mark. size:{} lowWaterMark:{} highWaterMark:{} maxSize:{}",
+                mName, mMap.size(), mLowWaterMark, mHighWaterMark, mMaxSize);
+            mNextAllowedSizeWarning = now.plus(mWarnInterval);
+          }
+        }
+        evictionCount += evictBatch(Math.min(mMap.size() - mLowWaterMark, mEvictBatchSize));
```

Comment:
could `mMap.size() - mLowWaterMark` ever be `< 0`?

Reply:
`evictBatch` handles non-positive values by doing nothing